File: backend/rojgarnext/app/core/location/distance_calculator.py
# app/core/location/distance_calculator.py
# PYTHON VERSION - BACKEND KE LIYE

import math

class DistanceCalculator:
    EARTH_RADIUS_KM = 6371.0
    EARTH_RADIUS_METERS = 6371000.0

    @staticmethod
    def calculate_distance(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
        """
        दो coordinates के बीच हवाई दूरी निकाले (मीटर में)
        """
        lat1_rad = math.radians(lat1)
        lat2_rad = math.radians(lat2)
        delta_lat = math.radians(lat2 - lat1)
        delta_lon = math.radians(lon2 - lon1)

        a = math.sin(delta_lat / 2) ** 2 + \
            math.cos(lat1_rad) * math.cos(lat2_rad) * \
            math.sin(delta_lon / 2) ** 2
        c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))

        return DistanceCalculator.EARTH_RADIUS_METERS * c

    @staticmethod
    def calculate_distance_km(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
        """दूरी किलोमीटर में"""
        distance_m = DistanceCalculator.calculate_distance(lat1, lon1, lat2, lon2)
        return distance_m / 1000.0

    @staticmethod
    def format_distance(lat1: float, lon1: float, lat2: float, lon2: float) -> str:
        """सुंदर फॉर्मेट में दूरी"""
        distance_m = DistanceCalculator.calculate_distance(lat1, lon1, lat2, lon2)
        if distance_m >= 1000:
            return f"{distance_m / 1000:.2f} km"
        else:
            return f"{distance_m:.0f} m"

    @staticmethod
    def is_within_distance(lat1: float, lon1: float, lat2: float, lon2: float, threshold_meters: float = 100) -> bool:
        """चेक करे कि दो location threshold के अंदर हैं या नहीं"""
        distance = DistanceCalculator.calculate_distance(lat1, lon1, lat2, lon2)
        return distance <= threshold_meters

File: frontend/rojgarnext/lib/core/utils/distance_calculator.dart
// lib/core/utils/distance_calculator.dart
// COMPLETE DISTANCE CALCULATION BETWEEN TWO COORDINATES

import 'dart:math';

class DistanceCalculator {
  static const double earthRadiusMeters = 6371000.0;
  static const double earthRadiusKm = 6371.0;

  static double calculateDistance(
    double lat1,
    double lon1,
    double lat2,
    double lon2,
  ) {
    final lat1Rad = _toRadians(lat1);
    final lat2Rad = _toRadians(lat2);
    final deltaLat = _toRadians(lat2 - lat1);
    final deltaLon = _toRadians(lon2 - lon1);

    final a = sin(deltaLat / 2) * sin(deltaLat / 2) +
        cos(lat1Rad) * cos(lat2Rad) * sin(deltaLon / 2) * sin(deltaLon / 2);
    final c = 2 * atan2(sqrt(a), sqrt(1 - a));

    return earthRadiusMeters * c;
  }

  static double calculateDistanceInKm(
    double lat1,
    double lon1,
    double lat2,
    double lon2,
  ) {
    return calculateDistance(lat1, lon1, lat2, lon2) / 1000.0;
  }

  static String formatDistance(
    double lat1,
    double lon1,
    double lat2,
    double lon2,
  ) {
    final distanceInMeters = calculateDistance(lat1, lon1, lat2, lon2);

    if (distanceInMeters >= 1000) {
      final km = distanceInMeters / 1000;
      return "${km.toStringAsFixed(2)} km";
    } else {
      return "${distanceInMeters.toStringAsFixed(0)} m";
    }
  }

  static bool isWithinDistance(
    double lat1,
    double lon1,
    double lat2,
    double lon2,
    double thresholdMeters,
  ) {
    return calculateDistance(lat1, lon1, lat2, lon2) <= thresholdMeters;
  }

  static bool areSameLocation(
    double lat1,
    double lon1,
    double lat2,
    double lon2,
  ) {
    return isWithinDistance(lat1, lon1, lat2, lon2, 100.0);
  }

  static double _toRadians(double degrees) {
    return degrees * pi / 180;
  }
}